{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "22abba08",
   "metadata": {},
   "source": [
    "# Preprocessing of TCR databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8695468",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import os\n",
    "import sys\n",
    "sys.path.append('../tcr_benchmark')\n",
    "\n",
    "from utils.config import col_va, col_ja, col_cdr3a, col_vb, col_jb, col_cdr3b, col_epitope, col_mhc\n",
    "from utils.config import required_cols\n",
    "\n",
    "sb.set_style('whitegrid')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0910f218",
   "metadata": {},
   "source": [
    "## Helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "30d0826a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IEDB does not save leading C and ending F or W in the sequence. We will therefore remove all of them to have it uniform.\n",
    "def trim_tcr(cdr3, is_beta):\n",
    "    if type(cdr3) != str:\n",
    "        return cdr3\n",
    "    idx_start = 0\n",
    "    if cdr3[0] == 'C':\n",
    "        idx_start = 1\n",
    "\n",
    "    idx_end = -1\n",
    "    if cdr3[-1] == 'F':\n",
    "        idx_end = -2\n",
    "    if cdr3[-1] == 'W' and not is_beta:\n",
    "        idx_end = -2\n",
    "    return cdr3[idx_start: idx_end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "59926166",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_base = '../tcr_benchmark/data'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12364207",
   "metadata": {},
   "source": [
    "## Databases\n",
    "- IEDB\n",
    "- VDJdb\n",
    "- McPas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2125cc48",
   "metadata": {},
   "source": [
    "### IEDB\n",
    "\n",
    "https://www.iedb.org/\n",
    "\n",
    "- Eptiope = Any\n",
    "- Assay = T Cell, Outcome = Positive\n",
    "- MHC Restriction = Any\n",
    "- Host = Any\n",
    "- Disease = Any\n",
    "\n",
    "=> Search => Receptors => Export Results => Export to CSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c636e013",
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_iedb = {\n",
    "    'Description': col_epitope,  \n",
    "\n",
    "    'Calculated V Gene': col_va,\n",
    "    'Calculated J Gene': col_ja,\n",
    "    'CDR3 Curated': col_cdr3a,\n",
    "\n",
    "    'Calculated V Gene.1' : col_vb,\n",
    "    'Calculated J Gene.1': col_jb,\n",
    "    'CDR3 Curated.1': col_cdr3b,\n",
    "    'MHC Allele Names': col_mhc,\n",
    "    'Name': col_epitope\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ac4a76d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\AppData\\Local\\Temp\\ipykernel_35996\\1959074555.py:2: DtypeWarning: Columns (2,10,12,13,14,15,16,17,18,19,20,21,22,23,24,29,30,35,36,41,42,43,44,45,46,47,48,49,50,51,52,53,58,59,64,65) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_iedb = pd.read_csv(path_iedb, skiprows=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3_alpha</th>\n",
       "      <th>V_alpha</th>\n",
       "      <th>J_alpha</th>\n",
       "      <th>CDR3_beta</th>\n",
       "      <th>V_beta</th>\n",
       "      <th>J_beta</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>MHC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IVVRSSNTGKL</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ37*01</td>\n",
       "      <td>ASSQDRDTQ</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>VMAPRTLIL</td>\n",
       "      <td>HLA-E*01:01, HLA-E*01:03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASSQGGDRGDPGDGY</td>\n",
       "      <td>TRBV5-6</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "      <td>CINGVCWTV</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AVTTDSWGKL</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ24*02</td>\n",
       "      <td>ASRPGLAGGRPEQ</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>LLFGYPVYV</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AVTTDSWGKL</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ24*02</td>\n",
       "      <td>ASRPGLMSAQPEQ</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>LLFGYPVYV</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AVRPTSGGSYIP</td>\n",
       "      <td>TRAV21*01</td>\n",
       "      <td>TRAJ6*01</td>\n",
       "      <td>ASSYVGNTGEL</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>TRBJ2-2*01</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     CDR3_alpha      V_alpha    J_alpha        CDR3_beta      V_beta  \\\n",
       "0   IVVRSSNTGKL  TRAV26-1*01  TRAJ37*01        ASSQDRDTQ   TRBV14*01   \n",
       "1           NaN          NaN        NaN  ASSQGGDRGDPGDGY     TRBV5-6   \n",
       "2    AVTTDSWGKL  TRAV12-2*01  TRAJ24*02    ASRPGLAGGRPEQ  TRBV6-5*01   \n",
       "3    AVTTDSWGKL  TRAV12-2*01  TRAJ24*02    ASRPGLMSAQPEQ  TRBV6-5*01   \n",
       "4  AVRPTSGGSYIP    TRAV21*01   TRAJ6*01      ASSYVGNTGEL  TRBV6-5*01   \n",
       "\n",
       "       J_beta    Epitope                       MHC  \n",
       "0  TRBJ2-3*01  VMAPRTLIL  HLA-E*01:01, HLA-E*01:03  \n",
       "1     TRBJ1-2  CINGVCWTV               HLA-A*02:01  \n",
       "2  TRBJ2-7*01  LLFGYPVYV               HLA-A*02:01  \n",
       "3  TRBJ2-7*01  LLFGYPVYV               HLA-A*02:01  \n",
       "4  TRBJ2-2*01  SLLMWITQC               HLA-A*02:01  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_iedb = '../data/raw/receptor_table_export_1698079287.csv'\n",
    "df_iedb = pd.read_csv(path_iedb, skiprows=1)\n",
    "\n",
    "df_iedb = df_iedb.rename(columns=rename_iedb)\n",
    "df_iedb = df_iedb[required_cols]\n",
    "\n",
    "df_iedb = df_iedb[~df_iedb[col_epitope].isna()]\n",
    "df_iedb = df_iedb[~(df_iedb[col_cdr3a].isna() & df_iedb[col_cdr3b].isna())]\n",
    "\n",
    "df_iedb[col_cdr3b] = df_iedb[col_cdr3b].apply(lambda x: trim_tcr(x, True))\n",
    "df_iedb[col_cdr3a] = df_iedb[col_cdr3a].apply(lambda x: trim_tcr(x, False))\n",
    "\n",
    "df_iedb = df_iedb.drop_duplicates([col_epitope, col_cdr3a, col_cdr3b])\n",
    "df_iedb = df_iedb.reset_index(drop=True)\n",
    "\n",
    "df_iedb.to_csv('../data/processed/iedb.csv')\n",
    "df_iedb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a2c5833b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Epitopes:  2623\n",
      "Unique TCRs:  148878\n",
      "Total pairs:  191170\n"
     ]
    }
   ],
   "source": [
    "print('Unique Epitopes: ', len(df_iedb[col_epitope].unique()))\n",
    "print('Unique TCRs: ', len(df_iedb[col_cdr3b].unique()))\n",
    "print('Total pairs: ', len(df_iedb))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6edc92dd",
   "metadata": {},
   "source": [
    "## VDJdb\n",
    "\n",
    "https://github.com/antigenomics/vdjdb-db/releases/download/2022-03-30/vdjdb-2022-03-30.zip\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "974a4419",
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_vdjdb = {\n",
    "    'antigen.epitope': col_epitope, \n",
    "\n",
    "    'v.segm_a': col_va,\n",
    "    'j.segm_a': col_ja,\n",
    "    'cdr3_a': col_cdr3a,\n",
    "\n",
    "    'v.segm_b' : col_vb,\n",
    "    'j.segm_b': col_jb,\n",
    "    'cdr3_b': col_cdr3b,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4f090c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_unique_index(x):\n",
    "    idcs = []\n",
    "    next_idx = x.max() + 1\n",
    "    for i in x:\n",
    "        if i != 0:\n",
    "            idcs.append(i)\n",
    "        else:\n",
    "            idcs.append(next_idx)\n",
    "            next_idx += 1\n",
    "    return idcs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0a7a1072",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3_alpha</th>\n",
       "      <th>V_alpha</th>\n",
       "      <th>J_alpha</th>\n",
       "      <th>CDR3_beta</th>\n",
       "      <th>V_beta</th>\n",
       "      <th>J_beta</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>MHC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IVRAPGRADM</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>ASSYLPGQGDHYSNQPQ</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASSFEAGQGFFSNQPQ</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AVPSGAGSYQL</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>ASSFEPGQGFYSNQPQ</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AVKASGSRL</td>\n",
       "      <td>TRAV2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASSYEPGQVSHYSNQPQ</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AYRPPGTYKY</td>\n",
       "      <td>TRAV38-2/DV8*01</td>\n",
       "      <td>TRAJ40*01</td>\n",
       "      <td>ASSALASLNEQ</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>FLKEKGGL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CDR3_alpha          V_alpha    J_alpha          CDR3_beta     V_beta  \\\n",
       "0   IVRAPGRADM      TRAV26-1*01  TRAJ43*01  ASSYLPGQGDHYSNQPQ  TRBV13*01   \n",
       "1          NaN              NaN        NaN   ASSFEAGQGFFSNQPQ  TRBV13*01   \n",
       "2  AVPSGAGSYQL        TRAV20*01  TRAJ28*01   ASSFEPGQGFYSNQPQ  TRBV13*01   \n",
       "3    AVKASGSRL         TRAV2*01        NaN  ASSYEPGQVSHYSNQPQ  TRBV13*01   \n",
       "4   AYRPPGTYKY  TRAV38-2/DV8*01  TRAJ40*01        ASSALASLNEQ  TRBV14*01   \n",
       "\n",
       "       J_beta   Epitope  MHC  \n",
       "0  TRBJ1-5*01  FLKEKGGL  NaN  \n",
       "1  TRBJ1-5*01  FLKEKGGL  NaN  \n",
       "2  TRBJ1-5*01  FLKEKGGL  NaN  \n",
       "3  TRBJ1-5*01  FLKEKGGL  NaN  \n",
       "4  TRBJ2-1*01  FLKEKGGL  NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_vdjdb = '../data/raw/vdjdb-2023-06-01/vdjdb.txt'\n",
    "df_vdjdb = pd.read_csv(path_vdjdb, sep='\\t')\n",
    "\n",
    "df_vdjdb['complex.id'] = assign_unique_index(df_vdjdb['complex.id'])\n",
    "df_vdjdb = df_vdjdb.set_index('complex.id')\n",
    "\n",
    "df_vdjdb_a = df_vdjdb[df_vdjdb['gene']=='TRA'][['cdr3', 'v.segm', 'j.segm']]\n",
    "df_vdjdb_a.columns = [f'{c}_a' if c in ['cdr3', 'v.segm', 'j.segm'] else c for c in df_vdjdb_a.columns]\n",
    "\n",
    "df_vdjdb_b = df_vdjdb[df_vdjdb['gene']=='TRB'][['cdr3', 'v.segm', 'j.segm']]\n",
    "df_vdjdb_b.columns = [f'{c}_b' if c in ['cdr3', 'v.segm', 'j.segm'] else c for c in df_vdjdb_b.columns]\n",
    "\n",
    "df_vdjdb = df_vdjdb[~df_vdjdb.index.duplicated(keep='first')]\n",
    "df_vdjdb = pd.concat([df_vdjdb, df_vdjdb_a, df_vdjdb_b], axis=1)\n",
    "\n",
    "df_vdjdb = df_vdjdb.rename(columns=rename_vdjdb)\n",
    "df_vdjdb['MHC'] = np.nan\n",
    "df_vdjdb = df_vdjdb[required_cols]\n",
    "\n",
    "df_vdjdb[col_cdr3b] = df_vdjdb[col_cdr3b].apply(lambda x: trim_tcr(x, True))\n",
    "df_vdjdb[col_cdr3a] = df_vdjdb[col_cdr3a].apply(lambda x: trim_tcr(x, False))\n",
    "\n",
    "df_vdjdb = df_vdjdb.reset_index(drop=True)\n",
    "\n",
    "df_vdjdb.to_csv('../data/processed/vdjdb.csv')\n",
    "df_vdjdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "264341eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Epitopes:  1169\n",
      "Unique TCRs:  41727\n",
      "Total pairs:  62177\n"
     ]
    }
   ],
   "source": [
    "print('Unique Epitopes: ', len(df_vdjdb[col_epitope].unique()))\n",
    "print('Unique TCRs: ', len(df_vdjdb[col_cdr3b].unique()))\n",
    "print('Total pairs: ', len(df_vdjdb))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7e11da2",
   "metadata": {},
   "source": [
    "## McPAS-TCR\n",
    "http://friedmanlab.weizmann.ac.il/McPAS-TCR/\n",
    "\n",
    "\"Download the complete database\" => downloaded 23.10.2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f5be0da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_mcpas = {\n",
    "    'Epitope.peptide': col_epitope, \n",
    "\n",
    "    'TRAV': col_va,\n",
    "    'TRAJ': col_ja,\n",
    "    'CDR3.alpha.aa': col_cdr3a,\n",
    "\n",
    "    'TRBV' : col_vb,\n",
    "    'TRBJ': col_jb,\n",
    "    'CDR3.beta.aa': col_cdr3b,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d1778bc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\AppData\\Local\\Temp\\ipykernel_35996\\2531970576.py:2: DtypeWarning: Columns (13,17,24,26,28) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_mcpas = pd.read_csv(path_mcpas, encoding = \"ISO-8859-1\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Epitope</th>\n",
       "      <th>V_alpha</th>\n",
       "      <th>J_alpha</th>\n",
       "      <th>CDR3_alpha</th>\n",
       "      <th>V_beta</th>\n",
       "      <th>J_beta</th>\n",
       "      <th>CDR3_beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IKAVYNFATCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV8-1</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>ASSDAGANTE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IKAVYNFATCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV8-1</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>ASSDAGAYAE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IKAVYNFATCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV8-3</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>ASSDAGGAAE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IKAVYNFATCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV8-1</td>\n",
       "      <td>TRBJ1-6</td>\n",
       "      <td>ASSDAGHSPL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IKAVYNFATCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV8-3</td>\n",
       "      <td>TRBJ2-6</td>\n",
       "      <td>ASSDAWGGAEQ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Epitope V_alpha J_alpha CDR3_alpha   V_beta   J_beta    CDR3_beta\n",
       "0  IKAVYNFATCG     NaN     NaN        NaN  TRBV8-1  TRBJ1-1   ASSDAGANTE\n",
       "1  IKAVYNFATCG     NaN     NaN        NaN  TRBV8-1  TRBJ2-1   ASSDAGAYAE\n",
       "2  IKAVYNFATCG     NaN     NaN        NaN  TRBV8-3  TRBJ1-1   ASSDAGGAAE\n",
       "3  IKAVYNFATCG     NaN     NaN        NaN  TRBV8-1  TRBJ1-6   ASSDAGHSPL\n",
       "4  IKAVYNFATCG     NaN     NaN        NaN  TRBV8-3  TRBJ2-6  ASSDAWGGAEQ"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_mcpas = '../data/raw/McPAS-TCR.csv'\n",
    "df_mcpas = pd.read_csv(path_mcpas, encoding = \"ISO-8859-1\")\n",
    "df_mcpas = df_mcpas[rename_mcpas.keys()]\n",
    "df_mcpas = df_mcpas.rename(columns=rename_mcpas)\n",
    "\n",
    "df_mcpas[col_cdr3b] = df_mcpas[col_cdr3b].apply(lambda x: trim_tcr(x, True))\n",
    "df_mcpas[col_cdr3a] = df_mcpas[col_cdr3a].apply(lambda x: trim_tcr(x, False))\n",
    "\n",
    "df_mcpas = df_mcpas.reset_index(drop=True)\n",
    "\n",
    "df_mcpas.to_csv('../data/processed/mcpas_tcr.csv')\n",
    "df_mcpas.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e05e8108",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Epitopes:  380\n",
      "Unique TCRs:  31201\n",
      "Total pairs:  39985\n"
     ]
    }
   ],
   "source": [
    "print('Unique Epitopes: ', len(df_mcpas[col_epitope].unique()))\n",
    "print('Unique TCRs: ', len(df_mcpas[col_cdr3b].unique()))\n",
    "print('Total pairs: ', len(df_mcpas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4308dcb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:benchmarkPlotting]",
   "language": "python",
   "name": "conda-env-benchmarkPlotting-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
